### PR TITLE
Blazor Prevent Stale Connection Usage Upon Reconnect

### LINKED ISSUE #40608: Blazor server - Using IJSStreamReference after browser reconnect throws "Cannot send data if the connection is not in the 'Connected' State"
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

In our application we use `IJSStreamReference`s to transfer data (Blobs) from the browser to the server. We've found the following bug in that functionality:

When you've used a `IJSStreamReference` in the current Blazor session, let the browser lose its connection (so the reconnect message is shown), reconnect again and use a (new) `IJSStreamReference`, you'll get the following error in the browser: `Cannot send data if the connection is not in the 'Connected' State`. And after a while (30 seconds or so?) you'll get a `TimeoutException` at the server.

Note that we don't reuse the `IJSStreamReference` and put the correct `using` statements in the code. See the Steps To Reproduce for the link to a small sample project.

### Expected Behavior

`IJSStreamReference` can be used after the browser has reconnected to its Blazor session.

### Steps To Reproduce

Link to a sample project to reproduce the bug: https://github.com/Mrlten/BlazorIJSStreamReferenceError

All functionality for the repro can be found in the `site.js`, `Index.razor` and `Index.razor.cs`.

Steps to reproduce:
1. Start the sample application.
2. Click on the 'Retrieve data using IJSStreamReference' button at least once (the `IJSStreamReference` must be used at least once in the current session to get the error).
3. Let the browser lose its connection to the server (the browser must connect to the original session on the server so don't shutdown the server process).
4. After the browser reconnected to the (original) session click on the 'Retrieve data using IJSStreamReference' button again.
5. Now the following things will happen:
   - The `RetrieveBlobAsync` server-side method is succesfully called.
   - The `retrieveBlob` javascript function is succesfully called.
   - The following javascript error is immediately thrown in the browser's console:
```
Uncaught (in promise) Error: Cannot send data if the connection is not in the 'Connected' State.
    at bt.send (blazor.server.js:1:52004)
    at kt._sendMessage (blazor.server.js:1:68639)
    at kt._sendWithProtocol (blazor.server.js:1:68680)
    at kt.send (blazor.server.js:1:68788)
    at blazor.server.js:1:129930
````
   - After 30 (?) seconds the following error is thrown at the server process:
```
fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
    Unhandled exception in circuit 'gdrrToRh8Rwr9lh70UeQR8_kBwaW5nkxAtDwy4y1LYo'.
    System.TimeoutException: Did not receive any data in the allotted time.
    at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
    at System.IO.Pipelines.Pipe.GetReadAsyncResult()
    at System.IO.Pipelines.PipeReaderStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
```
   - Followed by two `Microsoft.JSInterop.JSDisconnectedException`'s.

### Exceptions (if any)

In the browser:
```
Uncaught (in promise) Error: Cannot send data if the connection is not in the 'Connected' State.
    at bt.send (blazor.server.js:1:52004)
    at kt._sendMessage (blazor.server.js:1:68639)
    at kt._sendWithProtocol (blazor.server.js:1:68680)
    at kt.send (blazor.server.js:1:68788)
    at blazor.server.js:1:129930
```

At the server after a while:
```
fail: Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost[111]
    Unhandled exception in circuit 'gdrrToRh8Rwr9lh70UeQR8_kBwaW5nkxAtDwy4y1LYo'.
    System.TimeoutException: Did not receive any data in the allotted time.
    at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
    at System.IO.Pipelines.Pipe.GetReadAsyncResult()
    at System.IO.Pipelines.PipeReaderStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
```

### .NET Version

6.0.201

### Anything else?

 - Not dependent on the VS version
 - dotnet info output:
```
.NET SDK (reflecting any global.json):
 Version:   6.0.201
 Commit:    ef40e6aa06

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.201\

Host (useful for support):
  Version: 6.0.3
  Commit:  c24d9a9c91

.NET SDKs installed:
  6.0.201 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### PR DESCRIPTION
**Pull request by TanayParikh**
As @Mrlten pointed out in the original issue:

> Even though a new connection is created including a new `sendJSDataStream` handler the old handler is being called. This is because the `sendJSDataStream` function is cached by the [DotNet.JSObject class](https://github.com/dotnet/aspnetcore/blob/main/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L22).

Thus, the original `sendJSDataStream` is cached, with the original `connection` in an inline lambda function. This prevents the function from using the **new** `connection` upon reconnection. This is why upon reconnection, you get an error about `Cannot send data if the connection is not in the 'Connected' State`, it's still using the old connection. 

I've updated our connection initialization to use a global `_connection` property rather than the local connection variable within the function. This ensures cached functions always utilize the _current_ `connection`, rather than caching the `connection` from the first call to the function.

While I haven't specifically tried repro'ing the issue with other framework functions which use a (cached) lambda with the function local `connection` it's possible they may be facing the same issue. Regardless, the fix in this PR will resolve that issue if it is in fact present. 

Fixes: https://github.com/dotnet/aspnetcore/issues/40608

### CODE REVIEW

File: src/Components/Web.JS/src/Boot.Server.ts [hunk 1]
```diff
@@ -21,6 +21,8 @@ import { fetchAndInvokeInitializers } from './JSInitializers/JSInitializers.Serv
 let renderingFailed = false;
 let started = false;
 
+let _connection: HubConnection;
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Boot.Server.ts:24]: We don't have an underscore prefix convention for the other variables in this file. Could we change it to `connection` for consistency?

I'm glad you did call it `_connection` for the initial review though as it makes it clearer which other usage sites are affected 😄

**Review by javiercn — APPROVED** [review]
Agree with @SteveSandersonMS other than that, it looks good.

**Comment by SteveSandersonMS** [comment]
While I see that this change will fix the immediate issue, I wonder if we could tidy up the code a bit further. My concern is that we're taking some confusing code that was subtly broken, and replacing it with differently-confusing code that might be subtly broken in different ways (now or in the future).

Previously, we didn't have any global `connection` because on each reconnection, we're re-run `initializeConnection` and it would hold its own `connection` object and use it in a newly-created set of closures that were wired up to events for the new connection. Generally this worked **except** for the specific case of:

 * Assigning one of these closures to a global (`Blazor._internal.forceCloseConnection` or `Blazor._internal.sendJSDataStream`)
 * ... then calling that global via JS interop, which caches the name->function_instance mapping, and hence only sees the first such assigned instance

The change in this PR avoids the immediate problem by letting the closure continue to be cached, but having it refer to a global mutable `connection` variable instead of capturing one from `initializeConnection`. That works but still has issues:

 * We're still caching and reusing an arbitrary "first" instance of the function. The logic makes it look like the function instance could be replaced by a new one, but actually the cached value will be used, so that's potentially misleading.
 * Logic within `initializeConnection` might be trying to do things with the old and replaced `connection` instance, for example to fully shut it down after a new one has been established. But any such logic would be problematic because it would start acting on the new connection instance.

One way we could untangle all of this is to really try to stop overwriting function instances on `Blazor._internal.*`. I've always been uncomfortable with how we do that. If instead, `Blazor._internal.sendJSDataStream` was defined once up front like a regular function and not dynamically assigned inside `initializeConnection`, then there would be no future confusion about caching. Its internal logic will need to refer to some mutable `currentConnection` instance which we'd overwrite inside `initializeConnection`, but the rest of `initializeConnection`'s logic is free to refer to either the global "current connection" or its own captured "my connection which may or may not still be current" - and in most cases it does want to refer to its own connection, as there's no value in making the old handlers start acting on the new conneciton, as that clashes with newer handlers that do act on the new connection.

Hopefully that would not be a big refactoring, but if we can avoid overwriting `Blazor._internal.sendJSDataStream`/`forceCloseConnection` then this kind of cache-based confusion can no longer exist. What do you think?

**Comment by TanayParikh** [comment]
> Hopefully that would not be a big refactoring, but if we can avoid overwriting `Blazor._internal.sendJSDataStream`/`forceCloseConnection` then this kind of cache-based confusion can no longer exist. What do you think?

Thanks for the feedback, I believe https://github.com/dotnet/aspnetcore/pull/40702/commits/480cc8fd7798c7390a7d3fe94bbd2985e71b62c6 should address this now.

**Comment by TanayParikh** [comment]
/ping @SteveSandersonMS is this good to go now?

**Review by SteveSandersonMS — APPROVED** [review]
Great, thanks for the update!

**Pull request merged**